ksql-engine/src/test/java/io/confluent/ksql/EndToEndEngineTest.java

```diff
@@ -0,0 +1,554 @@
+    put("auto.offset.reset", "earliest");
+    put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+  }};
+  private final Properties streamsProperties = new Properties();
+  private SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
+
+  private Query query;
+
+  private KsqlEngine ksqlEngine;
+
+  @Before
+  public void before() {
+    streamsProperties.putAll(config);
+    ksqlEngine = new KsqlEngine(
+        new KsqlConfig(config),
+        new FakeKafkaTopicClient(),
+        schemaRegistryClient,
+        metaStore);
+  }
+
+  @After
+  public void cleanUp() {
+    ksqlEngine.close();
+  }
+
+  protected static class Window {
+    private final long start;
+    private final long end;
+
+    public Window(long start, long end) {
+      this.start = start;
+      this.end = end;
+    }
+
+    public long size() {
+      return end - start;
+    }
+  }
+
+  protected interface SerdeSupplier<T> {
+    Serializer<T> getSerializer(SchemaRegistryClient schemaRegistryClient);
+    Deserializer<T> getDeserializer(SchemaRegistryClient schemaRegistryClient);
+  }
+
+  protected static class StringSerdeSupplier implements SerdeSupplier<String> {
+    @Override
+    public Serializer<String> getSerializer(SchemaRegistryClient schemaRegistryClient) {
+      return Serdes.String().serializer();
+    }
+
+    @Override
+    public Deserializer<String> getDeserializer(SchemaRegistryClient schemaRegistryClient) {
+      return Serdes.String().deserializer();
+    }
+  }
+
+  protected static class AvroSerdeSupplier implements SerdeSupplier {
+    @Override
+    public Serializer getSerializer(SchemaRegistryClient schemaRegistryClient) {
+      return new KafkaAvroSerializer(schemaRegistryClient);
+    }
+
+    @Override
+    public Deserializer getDeserializer(SchemaRegistryClient schemaRegistryClient) {
+      return new KafkaAvroDeserializer(schemaRegistryClient);
+    }
+  }
+
+  protected static class ValueSpecAvroDeserializer implements Deserializer<Object> {
+    private final SchemaRegistryClient schemaRegistryClient;
+    private final KafkaAvroDeserializer avroDeserializer;
+
+    public ValueSpecAvroDeserializer(SchemaRegistryClient schemaRegistryClient) {
+      this.schemaRegistryClient = schemaRegistryClient;
+      this.avroDeserializer = new KafkaAvroDeserializer(schemaRegistryClient);
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public void configure(Map<String, ?> properties, boolean b) {
+    }
+
+    @Override
+    public Object deserialize(String topicName, byte[] data) {
+      Object avroObject = avroDeserializer.deserialize(topicName, data);
+      String schemaString;
+      try {
+        schemaString = schemaRegistryClient.getLatestSchemaMetadata(topicName + "-value").getSchema();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      new org.apache.avro.Schema.Parser().parse(schemaString);
```

Comment:
not sure what the purpose of this line is?

Reply:
copypasta